### PR TITLE
Implement infinite scroll in dashboard

### LINKED ISSUE #493: Replace default time window with last N drives
connect currently only shows the last two weeks of drives by default. Switch it to show the last N drives instead (and load more once you hit the bottom). The date range should still be an input, just an optional one.

### PR DESCRIPTION
**Pull request by yaodingyd**
Fix https://github.com/commaai/connect/issues/493

https://github.com/commaai/connect/assets/11392695/1dbe06bb-95df-4df4-a85a-b2e4e88e41ba

The console shows we keep fetching data until found; and it will fetch new data when user scrolls to the bottom.

There are some question remainng to be answered:
1. I use a heuristic to do look back with some fallback. we might want to change that
2. the option for time filer is not fixed yet. what should it be called? "Recent Rides"?
3. I changed how the loading state display: when we have data and there's new data fetching, keep displaying current data and have a loader spiiner at the bottom.

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -23,4 +23,4 @@ jobs:
           path: ./dist
           overwrite: true
           retention-days: 1
-          name: build-artifacts-${{ github.event.number }}
+          name: build-artifacts-${{ github.run_id }}
```
Comment (adeebshihadeh) [inline .github/workflows/build.yaml]: looks like a bad merge?

File: .github/workflows/preview.yaml [hunk 2]
```diff
@@ -19,7 +19,7 @@ jobs:
       - name: Download build artifacts
         uses: actions/download-artifact@v4
         with:
-          name: build-artifacts-${{ github.event.number }}
+          name: build-artifacts-${{ github.event.workflow_run.id }}
```
Comment (adeebshihadeh) [inline .github/workflows/preview.yaml]: and here

**Review by adeebshihadeh — COMMENTED** [review]
Nice work!

1. The `routes_segments` endpoint accepts a `limit` parameter. Something like this should work: set `start` to the end of loaded routes, `end` to `start + 5 years`, and `limit` to 10. 
2. I think we want to get rid of the time filter on top, and make specifying the time range a more niche thing (small button next to "take snapshot?"). It's very rare that you'll want to specify a range.
3. Got a video of this?

**Comment by yaodingyd** [comment]
thanks for the review, will update per comment.

I've attached a short video in description, what information do you need more? I'll make a new one

**Comment by yaodingyd** [comment]
> The routes_segments endpoint accepts a limit parameter. Something like this should work: set start to the end of loaded routes, end to start + 5 years, and limit to 10.

I'm not sure if that would work. route_segments doesn't have this `limit` param:
https://github.com/commaai/comma-api/blob/6658a0e71e29ba6cfdb380f8b058ff91cd834719/src/drives.js#L13-L15

I see listRoutes has it
https://github.com/commaai/comma-api/blob/6658a0e71e29ba6cfdb380f8b058ff91cd834719/src/drives.js#L45-L51

but it doesn't return what i need

**Comment by yaodingyd** [comment]
I replaced the time select dropdown with a button you suggested:


https://github.com/commaai/connect/assets/11392695/bbc30d59-b8eb-4875-a7d4-6cea90665029

**Comment by adeebshihadeh** [comment]
> I'm not sure if that would work. route_segments doesn't have this `limit` param: [commaai/comma-api@`6658a0e`/src/drives.js#L13-L15](https://github.com/commaai/comma-api/blob/6658a0e71e29ba6cfdb380f8b058ff91cd834719/src/drives.js#L13-L15)

Ah that's incomplete. I was looking at the backend server code. If you add `limit`, it should work.

**Comment by yaodingyd** [comment]
The api module is a npm package. We would need to update it first though

**Comment by adeebshihadeh** [comment]
#503

**Comment by yaodingyd** [comment]
I've updated the implementation for infinite scrolling, it would use `routes_segment` with incremental limit:


https://github.com/commaai/connect/assets/11392695/a0f6afc7-d8e0-436d-a388-4ab3575c726a

I only use `3` as incremental value since there are not that many routes for demo account, and we can adjust this value


The manual time select still works

https://github.com/commaai/connect/assets/11392695/de3c643c-fcc2-4d69-8f7a-2d7c618a35eb

**Comment by adeebshihadeh** [comment]
I think 5 is a good increment. As a side effect, this should be a nice speedup for users who drive a lot!

Deployed preview doesn't work, but it still looks a bit buggy in your video. Let me know when it's ready for a proper review.

**Comment by yaodingyd** [comment]
I think it's good to review, preview is good to test


https://github.com/commaai/connect/assets/11392695/4931e00d-630d-4ffa-9a04-0211f15eb32b

File: src/actions/index.js [hunk 3]
```diff
@@ -83,11 +87,38 @@ export function checkRoutesData() {
...
+  };
+}
+
+export function checkLastRoutesData() {
+  return (dispatch, getState) => {
+    const limit = getState().limit
+    console.log(`fetching ${limit +LIMIT_INCREMENT } routes`)
+    dispatch({
+      type: Types.ACTION_UPDATE_ROUTE_LIMIT,
+      limit: limit + LIMIT_INCREMENT,
```
Comment (adeebshihadeh) [inline src/actions/index.js:118]: I don't think it should increment every request. I think we should only load `LIMIT_INCREMENT` new routes each time you reach the bottom of the list

File: src/components/explorer.jsx [hunk 4]
```diff
@@ -149,11 +149,10 @@ class ExplorerApp extends Component {
       this.props.dispatch(pause());
     }
 
-    // FIXME: ensures demo routes stay visible. can be removed once we're infinite scrolling
```
Comment (adeebshihadeh) [inline src/components/explorer.jsx:152]: this whole thing should be able to be deleted now

File: src/components/DeviceInfo/index.jsx [hunk 5]
```diff
@@ -511,6 +524,12 @@ class DeviceInfo extends Component {
             ? <CircularProgress size={ 19 } />
             : 'take snapshot'}
         </Button>
+        <Button
+          classes={{ root: `${classes.button} ${actionButtonClass}` }}
+          onClick={ this.onOpenTimeSelect }
+        >
+          set time filter
```
Comment (adeebshihadeh) [inline src/components/DeviceInfo/index.jsx]: let's make this some sort of clock icon

**Review by adeebshihadeh — COMMENTED** [review]
It's still quite broken if you just scroll up and down repeatedly. Maybe try setting the increment to 1 to emulate on the demo route?

image

File: src/components/explorer.jsx (hunk 4 above)
Comment (yaodingyd) [reply src/components/explorer.jsx:152]: added more comment about why it's necessary

File: src/actions/index.js (hunk 3 above)
Comment (yaodingyd) [reply src/actions/index.js:118]: I've updated functions called during init to avoid multiple calls to fetch routs

**Comment by yaodingyd** [comment]
mind taking another look?

**Comment by adeebshihadeh** [comment]
Still very broken. Immediately hit this on the demo:
image

**Comment by yaodingyd** [comment]
I fixed another bug related to time select component but I'm having a hard time to reproduce the bug in your screenshot. I tested the latest preview with different browsers too:

chrome

https://github.com/commaai/connect/assets/11392695/a8cc9ded-118a-495a-a6ba-33b7bc2aaef7

safari

https://github.com/commaai/connect/assets/11392695/9b96d3f6-97d2-443a-9724-fee409ac1ceb

All look good, both for just scrolling and using time select.

Could it be a blip on your machine?  Are you able to consistently reproduce the bug, and if so, could you give more details about your machine and browser?

**Comment by sshane** [comment]
I would add a little more padding on the bottom of "There are no more routes found in selected time range."

**Comment by yaodingyd** [comment]
@adeebshihadeh sorry for the ping, what's the next step here?

**Comment by adeebshihadeh** [comment]
Not sure if you fixed it, but I couldn't repro the bug when I went to screen record it. Can you rebase on the latest master?

**Comment by adeebshihadeh** [comment]
Instantly hit it again on Chrome on my Ubuntu machine. Here's a screen recording:

https://github.com/commaai/connect/assets/8762862/00c86f48-a3c9-4ef0-b7f9-a753aca277f5

**Comment by yaodingyd** [comment]
Finally figured out the issue: there is a race condition that the next page fetch would void the previous fetching in flight, like events and locations. I changed the code that we store any fetched event/location, and  update state would always try to update all these stored event/location.  Could you give it another try?

**Pull request merged**